java/fury-core/src/main/java/org/apache/fury/util/StringEncodingUtils.java

```diff
@@ -0,0 +1,362 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.fury.util;
+
+import static org.apache.fury.util.StringUtils.MULTI_CHARS_NON_LATIN_MASK;
+
+import org.apache.fury.memory.Platform;
+
+/** String Encoding Utils. */
+public class StringEncodingUtils {
+
+  public static int convertUTF16ToUTF8(char[] src, byte[] dst, int dp) {
+    int numChars = src.length;
+    for (int charOffset = 0; charOffset < numChars; ) {
+      if (charOffset + 4 <= numChars
+          && (Platform.getLong(src, Platform.CHAR_ARRAY_OFFSET + charOffset * 2L)
+                  & MULTI_CHARS_NON_LATIN_MASK)
+              == 0) {
+        // ascii only
+        dst[dp] = (byte) src[charOffset];
+        dst[dp + 1] = (byte) src[charOffset + 1];
+        dst[dp + 2] = (byte) src[charOffset + 2];
+        dst[dp + 3] = (byte) src[charOffset + 3];
+        dp += 4;
+        charOffset += 4;
+      } else {
+        char c = src[charOffset++];
+        if (c < 0x80) {
+          dst[dp++] = (byte) c;
+        } else if (c < 0x800) {
+          dst[dp] = (byte) (0xc0 | (c >> 6));
+          dst[dp + 1] = (byte) (0x80 | (c & 0x3f));
+          dp += 2;
+        } else if (c >= '\uD800' && c <= Character.MAX_LOW_SURROGATE) {
```

Comment:
How big `convertUTF16ToUTF8`?  Can this method be inlined to its caller?

Reply:
Since this method is called only when there are many ascii chars, this branch is less frequent. If this big method can't be inlined, we can move this branch into a new method.

Reply:
I'm not sure which will be faster, maybe need some benchmark. Could you use MediaContent for benchmark?

Reply:
This branch can be extracted, but I think for this kind of for loop with complex logic, iniline or not is not so critical, and even forcing inline of overly complex code can be a side effect.

Reply:
I see, thanks